### PR TITLE
remove connected disconnected state only

### PR DESCRIPTION
**Pull request by drahnr**
Reduces #3803 to the essential change-set to simplify review

Closes https://github.com/paritytech/polkadot/issues/3777

Renames:

* `RelayChainSelectionWithFallback` -> `RelayChainSelection`
* `RelayChainSelection` -> `RelayChainSelectionInner`

Adds:

* `OverseerConnector` usage
* `Basics`

Removes:

* `Overseer` `Connected`/`Disconnected` state handling and uses a `bool` `is_relay_chain` instead

### CODE REVIEW

**Review by ordian — APPROVED** [review]
Has been reviewed as part of #3803

File: node/overseer/overseer-gen/proc-macro/src/impl_builder.rs [hunk 1]
```diff
@@ -130,7 +130,7 @@ pub(crate) fn impl_builder(info: &OverseerInfo) -> proc_macro2::TokenStream {
 				&mut self.handle
 			}
 			/// Obtain access to the overseer handle.
-			pub fn as_handle(&mut self) -> &#handle {
```
Comment (Lldenaurois) [inline node/overseer/overseer-gen/proc-macro/src/impl_builder.rs:133]: Were these the causes of the warnings that would occur during macro expansion?

File: node/overseer/src/lib.rs [hunk 2]
```diff
@@ -142,18 +146,12 @@ where
...
-	Connected(OverseerHandle),
-}
+pub struct Handle(pub OverseerHandle);
 
 impl Handle {
-	/// Create a new disconnected [`Handle`].
-	pub fn new_disconnected() -> Self {
-		Self::Disconnected(Arc::new(RwLock::new(None)))
+	/// Create a new [`Handle`].
+	pub fn new(raw: OverseerHandle) -> Self {
```
Comment (Lldenaurois) [inline node/overseer/src/lib.rs:153]: I find a `impl From<OverseerHandle> for Handle` would be quite elegant here, but I can imagine we would want to be explicit when creating a handle.

File: node/overseer/src/lib.rs [hunk 3]
```diff
@@ -202,58 +200,8 @@ impl Handle {
...
-	/// after calling `connect_to_overseer` on `self` or a clone of `self`.
-	fn try_connect(&mut self) {
-		if let Self::Disconnected(ref mut x) = self {
-			let guard = x.write();
-			if let Some(ref h) = *guard {
-				let handle = h.clone();
-				drop(guard);
-				*self = Self::Connected(handle);
-			}
+		if self.0.send(event).await.is_err() {
```
Comment (Lldenaurois) [inline node/overseer/src/lib.rs:203]: So much better!

File: node/overseer/src/lib.rs [hunk 4]
```diff
@@ -680,7 +631,7 @@ where
 			.supports_parachains(supports_parachains)
 			.metrics(metrics.clone())
 			.spawner(s)
-			.build()?;
```
Comment (Lldenaurois) [inline node/overseer/src/lib.rs:683]: I feel like in this pattern, we may want to restructure as

```
.spawner(s)
.connector(connector)
.build()
```

**Review by Lldenaurois — APPROVED** [review]
I think there are a few places where we are explicit about the type, but needn't be,

e.g. let this: Option<_> = opt.map(...);

But I'm also not sure whether it's not better to be explicit in cases where you are returning an option or some other specific type.

I can see arguments for both, but I generally like to be as terse as possible. The type annotations are helpful to the reader, however.

File: node/overseer/overseer-gen/proc-macro/src/impl_builder.rs (hunk 1 above)
Comment (drahnr) [reply node/overseer/overseer-gen/proc-macro/src/impl_builder.rs:133]: This was just a borrow issue, the warnings with `+nightly` are partially resolved in #3874 , two remaining which are yet to be resolved.

File: node/overseer/src/lib.rs (hunk 4 above)
Comment (drahnr) [reply node/overseer/src/lib.rs:683]: `fn new` is going to die in #3874 :upside_down_face:

File: node/service/src/lib.rs [hunk 5]
```diff
@@ -1244,52 +1323,24 @@ pub fn new_chain_ops(
...
-				config,
-				jaeger_agent,
-				None,
-			)?;
-		return Ok((Arc::new(Client::Polkadot(client)), backend, import_queue, task_manager))
-	}
-
-	#[cfg(not(feature = "polkadot-native"))]
-	Err(Error::NoRuntime)
+	chain_ops!(config, jaeger_agent, telemetry_worker_handle; polkadot_runtime, PolkadotExecutorDispatch, Polkadot)
```
Comment (drahnr) [inline node/service/src/lib.rs:1343]:
```suggestion
	#[cfg(feature = "polkadot-native")]
	{
		chain_ops!(config, jaeger_agent, telemetry_worker_handle; polkadot_runtime, PolkadotExecutorDispatch, Polkadot)
	}
	
	#[cfg(not(feature = "polkadot-native"))]
	Err(Error::NoRuntime)
```

**Review by eskimor — APPROVED** [review]

File: node/service/src/lib.rs [hunk 6]
```diff
@@ -850,8 +914,10 @@ where
 		local_keystore.and_then(move |k| authority_discovery_service.map(|a| (a, k)));
 
 	let overseer_handle = if let Some((authority_discovery_service, keystore)) = maybe_params {
-		let (overseer, overseer_handle) = overseer_gen
+		// already have access to the handle
+		let (overseer, _handle) = overseer_gen
```
Comment (ordian) [inline node/service/src/lib.rs:918]: why is `_handle` here dropped? because `handle` is its clone?

**Pull request merged**